### PR TITLE
CI: Refactor ci reproducer

### LINKED ISSUE #36833: Docker reproducer for pipelines fails due to path mismatch
### Steps to reproduce

From a terminal:
```console
$ git clone https://github.com/spack/spack
$ git fetch origin pull/35450/merge:PR/35450
$ . spack/share/spack/setup-env.sh
$ spack ci reproduce-build https://gitlab.spack.io/api/v4/projects/2/jobs/6517196/artifacts --working-dir /tmp/pipeline
==> Fetching artifacts from: https://gitlab.spack.io/api/v4/projects/2/jobs/6517196/artifacts

==> Job ran with the following image: ghcr.io/spack/e4s-amazonlinux-2:v2023-03-09
checkout_commit: eecd2f984fe6eb5b3de6dc8af54a69a71684e02e
merge_commit: 554b6bb3998588978f28a903ad51d44932b90326
==> Job ran with the following tags: ['x86_64_v3', 'spack', 'public']

Run the following command:

    $ docker run --rm --name spack_reproducer -v /tmp/pipeline:/builds/spack/spack:Z -ti ghcr.io/spack/e4s-amazonlinux-2:v2023-03-09

Once inside the container:

    - Activate the environment

        $ source /builds/spack/spack/spack/share/spack/setup-env.sh
        $ spack env activate --without-view /builds/spack/spack/jobs_scratch_dir/concrete_environment

    - Run the install script

        $ /builds/spack/spack/jobs_scratch_dir/reproduction/install.sh

$ docker run --rm --name spack_reproducer -v /tmp/pipeline:/builds/spack/spack:Z -ti ghcr.io/spack/e4s-amazonlinux-2:v2023-03-09

```
Then within the docker container:
```console
bash-4.2# source /builds/spack/spack/spack/share/spack/setup-env.sh
bash-4.2# spack env activate --without-view /builds/spack/spack/jobs_scratch_dir/concrete_environment
```

### Error message

Trying to activate the environment in the container fails with:
```console
bash-4.2# spack env activate --without-view /builds/spack/spack/jobs_scratch_dir/concrete_environment
==> Error: Detected 3 missing include path(s):
   /builds/spack/spack/share/spack/gitlab/cloud_pipelines/configs
   /builds/spack/spack/share/spack/gitlab/cloud_pipelines/configs/linux
   /builds/spack/spack/share/spack/gitlab/cloud_pipelines/configs/linux/x86_64_v3
```
This is due to Spack being cloned in the wrong directory. As a workaround we can do:
```console
bash-4.2# ln -s /builds/spack/spack/spack/share /builds/spack/spack/share
```
to symlink the `share` directory where the environment expects it. The bug might be related to #34272

### Information on your system

* **Spack:** 0.20.0.dev0 (d918ae0bde903a2041a4c9d9edef5c59c63925cf)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by kwryankrattiger**
fixes #36833

* Autostart container
* Reproducer paths match CI paths
* Generate start scripts for docker and reproducer

@scottwittenburg

### CODE REVIEW

**Comment by kwryankrattiger** [comment]
@spackbot fix style

**Comment by kwryankrattiger** [comment]
@alalazo This should fix the container based reproducer and adds a couple nice quality of life features to make running the reproducer a little easier.

**Review by alecbcs — CHANGES_REQUESTED** [review]
@kwryankrattiger with the greater adoption of podman in HPC ecosystems, I think it'd be great if we could support both podman and docker here. The syntax for both is identical. You'd just need to detect if podman or docker is present on the system first.

**Comment by scottwittenburg** [comment]
> I think it'd be great if we could support both podman and docker here.

@alecbcs Supporting podman in general makes sense, but if the syntax is identical, why do we need to detect which is installed?  And do you see the command printing two sets of instructions in case both are installed?  I'm possibly just missing something since I don't have experience with podman.

At any rate, if this PR fixes broken behavior people are complaining about on slack, holding it up for podman support doesn't make sense to me.

**Comment by scottwittenburg** [comment]
@kwryankrattiger Do you have a link to a build that failed where we can try this out?  Or some other way to see that it's working?

**Comment by kwryankrattiger** [comment]
@alecbcs 
> with the greater adoption of podman in HPC ecosystems, I think it'd be great if we could support both podman and docker here. The syntax for both is identical. You'd just need to detect if podman or docker is present on the system first.

Since Gitlab CI isn't using podman and this is really needing to be merged, I don't think that is the highest priority. It could be a follow-on, I made an issue https://github.com/spack/spack/issues/38765.

In the case where you don't/can't have docker on the local system and need to run the reproducer, you can use `alias docker=podman` or install the [`podman-docker`](https://packages.fedoraproject.org/pkgs/podman/podman-docker/) package if you are using a RHEL like OS.

**Comment by kwryankrattiger** [comment]
@scottwittenburg 
> Do you have a link to a build that failed where we can try this out? Or some other way to see that it's working?

I tested this locally using an old download url from a previous reproducer. I tested the output with a rebuild command setup to fail and in the unit test.

File: lib/spack/spack/ci.py
Comment (alecbcs) [inline lib/spack/spack/ci.py]: @kwryankrattiger is there a use case why we'd not want to `set -e` when generating the script? E.g. why we'd want the script to continue executing even if commands start failing.

**Comment by alecbcs** [comment]
> if the syntax is identical, why do we need to detect which is installed? 

Although the syntax for launching a container is similar podman isn't often aliased by default to docker. So we'd need to know whether to generate a script with `docker CMD` or `podman  CMD`.

> And do you see the command printing two sets of instructions in case both are installed?

No I think that if docker is installed that could take a higher priority so we'd just generate one set of instructions.

> At any rate, if this PR fixes broken behavior people are complaining about on slack, holding it up for podman support doesn't make sense to me.

Sure, I didn't mean for this to be a large task. I was thinking if it'd help adoption of the command in more places, this could be as easy as generating the script with,

```
runtime="docker"
if type podman &>/dev/null; then
  runtime="podman"
fi

$runtime run --rm ...
```

or adding another config flag to the command,

```python
    reproduce.add_argument(
        "--runtime",
        help="Run producer with a given container engine",
        default="docker",
    )
```

**Comment by scottwittenburg** [comment]
@kwryankrattiger I just tried this to reproduce [this](https://gitlab.spack.io/spack/spack/-/jobs/7629661) build failure using your proposed changes in this PR.  It seems like it reproduced the error, but then the reproducer script died with:

```
Re-run install script using:
	/builds/spack/spack/jobs_scratch_dir/reproduction/install.sh
/reproducer/entrypoint.sh: 12: exec: : Permission denied
```

That wouldn't be the end of the world, but then it left me back at my terminal prompt instead of sitting in the container where I can start poking around to see what happened.

So my questions are: 

1. Was I supposed to use the `--interactive` argument in order to be sitting at the container prompt at the end?  If so, I kind of prefer to have that behavior as the default.

2. Is the `/reproducer/entrypoint.sh: 12: exec: : Permission denied` a problem you encountered?  Any idea what's going on there?

File: lib/spack/spack/ci.py
Comment (kwryankrattiger) [reply lib/spack/spack/ci.py]: I think I had a reason that was likely related to testing, but I don't have a note on it and I can't think of anything off hand why it should be guarded by run.

**Comment by kwryankrattiger** [comment]
> Was I supposed to use the --interactive argument in order to be sitting at the container prompt at the end? 

Yeah. I think maybe I should just all it "auto-start" and have the interactive portion be on all the time.

> Is the /reproducer/entrypoint.sh: 12: exec: : Permission denied a problem you encountered?

No, that part was working for me without issue. What are the permissions on the generated entrypoint.sh?

File: lib/spack/spack/ci.py
Comment (alecbcs) [reply lib/spack/spack/ci.py]: Gotcha, I'd argue we should keep `set -e` in the scripts regardless of run status.

Comment (kwryankrattiger) [reply lib/spack/spack/ci.py]: So last night while debugging a problem that @scottwittenburg ran into I remembered the reason for not including it. Running Docker with interactive requires the entrypoint script to not exit on failure, since it calls the `install.sh` script from the pipeline which we know is going to fail.

Comment (kwryankrattiger) [reply lib/spack/spack/ci.py]: I did change this to be more clear, since the logic that was there was in fact broken. Now, instead of `run` controlling that option, `exit_on_failure` was added to prevent early exit so `exec bash` can run after the install script failure.

**Comment by kwryankrattiger** [comment]
@alecbcs I added the runtime option. The issue with the if conditions is they are clunky in the process command routine, the idea is all of that logic is handled in python and the script just runs a pipeline.

**Comment by kwryankrattiger** [comment]
@spackbot fix style

**Comment by alecbcs** [comment]
@kwryankrattiger this is looking great! Thanks for adding the runtime option. Let me know when you're ready for a final review.

**Comment by kwryankrattiger** [comment]
@alecbcs @scottwittenburg  this should be ready for a final pass

File: lib/spack/spack/cmd/ci.py
Comment (alecbcs) [inline lib/spack/spack/cmd/ci.py]:
```suggestion
        help="Container runtime to use.",
```

Quick typo fix.

Comment (alecbcs) [reply lib/spack/spack/cmd/ci.py]: Otherwise this all looks good to me! I'll defer final merge decision to @scottwittenburg once the typo is fixed.

**Comment by scottwittenburg** [comment]
@kwryankrattiger Can you point me to the run that emitted the failure/reproducer message you used to test this most recently?  I can't get it to work, i.e. while it does reproduce the error, it still doesn't leave me at the container prompt.  Instead it prints:

```
Traceback (most recent call last):
  File "/builds/scott/pipeline-experiments/lib/spack/spack/main.py", line 1023, in main
    return _main(argv)
  File "/builds/scott/pipeline-experiments/lib/spack/spack/main.py", line 978, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/builds/scott/pipeline-experiments/lib/spack/spack/main.py", line 1006, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/builds/scott/pipeline-experiments/lib/spack/spack/main.py", line 647, in _invoke_command
    return_val = command(parser, args)
  File "/builds/scott/pipeline-experiments/lib/spack/spack/cmd/install.py", line 356, in install
    install_with_active_env(env, args, install_kwargs, reporter_factory)
  File "/builds/scott/pipeline-experiments/lib/spack/spack/cmd/install.py", line 425, in install_with_active_env
    env.install_specs(specs_to_install, **install_kwargs)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/builds/scott/pipeline-experiments/lib/spack/spack/report.py", line 251, in build_context_manager
    reporter.build_report(filename, specs=collector.specs)
  File "/builds/scott/pipeline-experiments/lib/spack/spack/reporters/cdash.py", line 264, in build_report
    self.finalize_report()
  File "/builds/scott/pipeline-experiments/lib/spack/spack/reporters/cdash.py", line 439, in finalize_report
    raise SpackError("Errors encountered, see above for more details")
spack.error.SpackError: Errors encountered, see above for more details
```

... and then returns me to my own terminal prompt.

**Comment by scottwittenburg** [comment]
I took your branch, added a WIP commit to break something, then made a merge commit like the sync script does to create this [job](https://gitlab.spack.io/scott/pipeline-experiments/-/jobs/7732269).  If you fetch from upstream spack and my fork, you should be able to try to reproduce that error and see if the same thing is happening to you.

**Comment by kwryankrattiger** [comment]
@spackbot fix style

**Review by scottwittenburg — APPROVED** [review]
Most recent testing worked!  When the reproducer hit the same problem as encountered in the cloud job, it left me at the container prompt as expected.

I also like having to copy-paste two fewer commands!  Thanks @kwryankrattiger!

**Comment by scottwittenburg** [comment]
@alecbcs Can you approve if you have no further change requests?  Thanks!

**Comment by alecbcs** [comment]
@kwryankrattiger @scottwittenburg is there any way we could get the unit tests to hit more of the reproducer code here?

**Comment by kwryankrattiger** [comment]
@alecbcs 
> @... is there any way we could get the unit tests to hit more of the reproducer code here?

For now I think the ratio of effort required to add those tests to added value of those tests is poor. This feature is currently broken on develop and holding it up in order to develop new testing infrastructure to test failure modes and gpg validation (which were both tested by @scottwittenburg and I extensively) is not a great plan.

I will make a follow on issue to create py-test mock infrastructure for a signed buildcaches and a spack repo so that those can be tested later on.

**Review by alecbcs — APPROVED** [review]
This is a good point, I think you're right it's worth merging this in to fix the currently broken functionality on develop. I suggest test cases as a way to prevent future headaches for you all  (so that rewrites to other parts of Spack don't unknowingly break this code requiring you to again go through and fix broken functionality), but if that's prioritized after this PR that's alright with me.

Side note: I find it a little jarring when folks enable auto-merge when PRs are waiting on a review since this'll automatically merge the second I hit approve. I'd prefer to instead hit approve, then hit merge and double check the merge commit message.

**Comment by haampie** [comment]
@kwryankrattiger can you improve the instructions?

```
==> Fetching artifacts from: https://gitlab.spack.io/api/v4/projects/2/jobs/8051851/artifacts

==> Job ran with the following image: ecpe4s/ubuntu20.04-runner-x86_64:2023-01-01
checkout_commit: 450235165955ebd939c1d25d98776e52ec5d620c
merge_commit: 7c784c3dff029e7454570853cf11cc59cc040be1
==> Job ran with the following tags: ['x86_64_v3', 'spack', 'small', 'public']
==> 
To run the docker reproducer:

    - Start the docker container install       $ /tmp/tmp.LdubPVrypg/ci_reproduction/start.sh
```

I have no clue what this is telling me to do.

**Comment by haampie** [comment]
Ah, "Start the docker container install" is not an instruction, but a description.

Why does it create  shell script?

**Comment by haampie** [comment]
It seems to run the build with only one job :thinking:

**Comment by haampie** [comment]
Wouldn't it be better to [execv](https://docs.python.org/3/library/os.html#os.execv) docker/podman instead of running it in a subprocess (with autorun)? Also wouldn't mind if autorun was the default.

**Comment by kwryankrattiger** [comment]
It should build with as many jobs as it did in CI, if not that is a bug I think.

It writes a script because the old way was just manually executing commands. The way it runs rhe script is because that is the tool I had from the existing CI module and I didn't see the need to reinvent the wheel for the reproducer. I think execv makes sense though since unlike other CI module tasks this really does takeover as the main program.

Autostart would be a convenient default, but I didn't want to do that since the old one gave users a chance to verify they put in the correct job and look at the artifacts first. I think if it is something people prefer though, default autostart sounds okay to me.

**Pull request merged**
